presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnwrapDateFunctionInPredicate.java

```diff
@@ -0,0 +1,249 @@
+            } catch (Exception e) {
+                System.out.println("Failed to parse VARCHAR as DateType: " + e.getMessage());
+                return Optional.empty();
+            }
+        }
+
+        // At this point the literal must be a DateType
+        if (!(literal.getType() instanceof DateType)) {
+            System.out.println("Literal type is still not DateType, exiting.");
+            return Optional.empty();
+        }
+
+        // Extract the epoch day from the literal (as a long)
+        Object value = literal.getValue();
+        long dateDays;
+        if (value instanceof Number) {
+            dateDays = ((Number) value).longValue();
+        }
+        else {
+            System.out.println("Unexpected literal value type: " + value.getClass());
+            return Optional.empty();
+        }
+
+        // Presto TIMESTAMP literals are stored as microseconds since the Unix epoch.
+        final long MICROSECONDS_PER_DAY = 86400000000L;
+        long lowerBoundMicros = dateDays * MICROSECONDS_PER_DAY;
+        long upperBoundMicros = (dateDays + 1) * MICROSECONDS_PER_DAY;
+
+        ConstantExpression lowerTimestampLiteral = new ConstantExpression(lowerBoundMicros, TimestampType.TIMESTAMP);
+        ConstantExpression upperTimestampLiteral = new ConstantExpression(upperBoundMicros, TimestampType.TIMESTAMP);
+
+        return Optional.of(createTimestampRangePredicate(column, lowerTimestampLiteral, upperTimestampLiteral));
+    }
+
+    private RowExpression createTimestampRangePredicate(
+            RowExpression column,
+            ConstantExpression lowerLiteral,
+            ConstantExpression upperLiteral)
+    {
+        // Build the comparisons: column >= lowerBound AND column < upperBound
+        CallExpression lowerBoundComparison = new CallExpression(
+                column.getSourceLocation(),
+                "GREATER_THAN_OR_EQUAL",
+                functionResolution.comparisonFunction(OperatorType.GREATER_THAN_OR_EQUAL, column.getType(), lowerLiteral.getType()),
+                BOOLEAN,
+                ImmutableList.of(column, lowerLiteral));
+
+        CallExpression upperBoundComparison = new CallExpression(
+                column.getSourceLocation(),
+                "LESS_THAN",
+                functionResolution.comparisonFunction(OperatorType.LESS_THAN, column.getType(), upperLiteral.getType()),
+                BOOLEAN,
+                ImmutableList.of(column, upperLiteral));
+
+        return new SpecialFormExpression(
+                column.getSourceLocation(),
+                AND,
+                BOOLEAN,
+                ImmutableList.of(lowerBoundComparison, upperBoundComparison));
+    }
+
+    /**
+     * Parses a date string in the format "yyyy-MM-dd" and returns the epoch day.
+     * (Epoch day = number of days since 1970-01-01)
+     */
+    private long parseEpochDay(String dateString)
```

Comment:
This should already be implemented somewhere which also accounts for leap years, etc. I don't think we need this function